### PR TITLE
Asphyxia core: init at 1.40d

### PR DESCRIPTION
**Pull request by ethernet8023**
###### Description of changes
Adding Asphyxia-core - a server for Konami rhythm games. https://asphyxia-core.github.io/
Asphyxia is closed-source, so this packages a binary. The license is set to `unfreeRedistributable`.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/doc/manual/release-notes/rl-2211.section.md [hunk 1]
```diff
@@ -23,12 +23,12 @@ In addition to numerous new and upgraded packages, this release has the followin
...
-   - `buildPlatform` is the platform that is responsible for building the NixOS
-     configuration. It defaults to the `hostPlatform`, for a non-cross
-     build configuration. To cross compile, set `buildPlatform` to a different
-     value.
+  - `hostPlatform` is the platform or "`system`" string of the NixOS system
+    described by the configuration.
+  - `buildPlatform` is the platform that is responsible for building the NixOS
+    configuration. It defaults to the `hostPlatform`, for a non-cross
+    build configuration. To cross compile, set `buildPlatform` to a different
+    value.
```
Comment (SuperSandro2000) [inline nixos/doc/manual/release-notes/rl-2211.section.md:31]: Please move all the unrelated changes into another PR.

File: pkgs/games/asphyxia-core/default.nix [hunk 2]
```diff
@@ -0,0 +1,75 @@
+with import <nixpkgs> { };
```
Comment (SuperSandro2000) [inline pkgs/games/asphyxia-core/default.nix:1]: Please use the default callPackage header.

File: pkgs/games/asphyxia-core/default.nix [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+stdenv.mkDerivation rec {
+  pname = "asphyxia-core";
+  version = "1.40";
+
+  src = fetchurl {
+    url = "https://github.com/asphyxia-core/asphyxia-core.github.io/releases/download/v${version}/asphyxia-core-linux-x64.zip";
```
Comment (SuperSandro2000) [inline pkgs/games/asphyxia-core/default.nix:8]: Can we build from source?

File: pkgs/games/asphyxia-core/default.nix [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    sha256 = "sha256-YFAXHzxoLuKfflwqYvpifefRE2/KbShuQ0v+0LG7OCg=";
+  };
+
+  dontStrip = true;
```
Comment (SuperSandro2000) [inline pkgs/games/asphyxia-core/default.nix:12]: Why?

File: pkgs/games/asphyxia-core/default.nix [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+
+  nativeBuildInputs = [ pkgs.unzip ];
```
Comment (SuperSandro2000) [inline pkgs/games/asphyxia-core/default.nix:14]:
```suggestion
  nativeBuildInputs = [ unzip ];
```

File: pkgs/games/asphyxia-core/default.nix [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+  buildInputs = [ stdenv.cc.cc.lib ];
+
+  sourceRoot = ".";
+
+  unpackPhase = ''
+    unzip $src
+  '';
```
Comment (SuperSandro2000) [inline pkgs/games/asphyxia-core/default.nix:21]: Isn't this done by default?

File: pkgs/games/asphyxia-core/default.nix [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+  buildPhase = ":";
```
Comment (SuperSandro2000) [inline pkgs/games/asphyxia-core/default.nix:22]:
```suggestion

  dontBuild = true;
```

File: pkgs/games/asphyxia-core/default.nix [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+
+  installPhase = ''
+    mkdir -p $out/bin
+    cp asphyxia-core $out/bin/asphyxia-core
```
Comment (SuperSandro2000) [inline pkgs/games/asphyxia-core/default.nix:26]:
```suggestion
    mv asphyxia-core $out/bin/asphyxia-core
```

File: pkgs/games/asphyxia-core/default.nix [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+  '';
+
+  preFixup = let
+    libPath = lib.makeLibraryPath [stdenv.cc.cc];
+  in ''
+    orig_size=$(stat --printf=%s $out/bin/asphyxia-core)
+    patchelf --set-interpreter "$(cat $NIX_CC/nix-support/dynamic-linker)" $out/bin/asphyxia-core
+    patchelf --set-rpath ${libPath} $out/bin/asphyxia-core
```
Comment (SuperSandro2000) [inline pkgs/games/asphyxia-core/default.nix:34]:
```suggestion
  preFixup = ''
    orig_size=$(stat --printf=%s $out/bin/asphyxia-core)
    patchelf --set-interpreter "$(cat $NIX_CC/nix-support/dynamic-linker)" $out/bin/asphyxia-core
    patchelf --set-rpath ${lib.makeLibraryPath [ stdenv.cc.cc ]} $out/bin/asphyxia-core
```

File: pkgs/games/asphyxia-core/default.nix [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+    description = "Asphyxia CORE";
+    homepage = "https://asphyxia-core.github.io/";
+    license = licenses.unfreeRedistributable;
+    platforms = [ "x86_64-linux" ];
+  };
+}
+
+
+
+
```
Comment (SuperSandro2000) [inline pkgs/games/asphyxia-core/default.nix:75]:
```suggestion
}
```

File: pkgs/games/asphyxia-core/default.nix [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+      value=$(dd if=$out/bin/asphyxia-core iflag=count_bytes,skip_bytes skip=$location \
+                 bs=1 count=$var_select status=none)
+      value=$(expr $shift_by + $value)
+      echo -n $value | dd of=$out/bin/asphyxia-core bs=1 seek=$location conv=notrunc
+    }
+    fix_offset PAYLOAD_POSITION
+    fix_offset PRELUDE_POSITION
+  '';
+  meta = with lib; {
+    description = "Asphyxia CORE";
```
Comment (SuperSandro2000) [inline pkgs/games/asphyxia-core/default.nix:66]: This should not just be the name of the program.

File: nixos/modules/services/games/asphyxia-core.nix [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+
+  configDir = pkgs.runCommand "createConfigDir" {} ''
+    mkdir -p $out
+    # create symlink to config file
+    ln -s ${configIniFile} $out/config.ini
+    cp -R ${corePlugins} $out/plugins
+  '';
+
+  startSh = pkgs.writeShellScript "asphyxia.sh" ''
+    CONFIG_DIR=${configDir}
```
Comment (SuperSandro2000) [inline nixos/modules/services/games/asphyxia-core.nix:31]: This should be set in the service via Environment

File: nixos/modules/services/games/asphyxia-core.nix [hunk 13]
```diff
@@ -0,0 +1,158 @@
...
+    echo "starting asphyxia in $CONFIG_DIR"
```
Comment (SuperSandro2000) [inline nixos/modules/services/games/asphyxia-core.nix:32]:
```suggestion
```

File: nixos/modules/services/games/asphyxia-core.nix [hunk 14]
```diff
@@ -0,0 +1,158 @@
...
+    exec -a $CONFIG_DIR/asphyxia-core ${cfg.package}/bin/asphyxia-core -d /var/lib/asphyxia-core
```
Comment (SuperSandro2000) [inline nixos/modules/services/games/asphyxia-core.nix:33]: This should be in ExecStart

File: nixos/modules/services/games/asphyxia-core.nix [hunk 15]
```diff
@@ -0,0 +1,158 @@
...
+    sections = {};
+  };
+
+  configIniFile = pkgs.writeText "config.ini" configAsIni;
+
+  configDir = pkgs.runCommand "createConfigDir" {} ''
+    mkdir -p $out
+    # create symlink to config file
+    ln -s ${configIniFile} $out/config.ini
+    cp -R ${corePlugins} $out/plugins
```
Comment (SuperSandro2000) [inline nixos/modules/services/games/asphyxia-core.nix:27]: Why do we need to cp here and can't symlink?

File: nixos/modules/services/games/asphyxia-core.nix [hunk 16]
```diff
@@ -0,0 +1,158 @@
...
+          enable_paseli = cfg.enablePaseli;
+    };
+    sections = {};
+  };
+
+  configIniFile = pkgs.writeText "config.ini" configAsIni;
+
+  configDir = pkgs.runCommand "createConfigDir" {} ''
+    mkdir -p $out
+    # create symlink to config file
```
Comment (SuperSandro2000) [inline nixos/modules/services/games/asphyxia-core.nix:25]:
```suggestion
```
that comment adds no value to the line below.

File: nixos/modules/services/games/asphyxia-core.nix [hunk 17]
```diff
@@ -0,0 +1,158 @@
...
+  configAsIni = generators.toINIWithGlobalSection {} {
+    globalSection = {
+          port = cfg.port;
+          bind = cfg.bind;
+          ping_ip = cfg.pingIp;
+          matching_port = cfg.matchingPort;
+          allow_register = cfg.allowRegister;
+          maintenance_mode = cfg.maintenanceMode;
+          enable_paseli = cfg.enablePaseli;
+    };
```
Comment (SuperSandro2000) [inline nixos/modules/services/games/asphyxia-core.nix:17]:
```suggestion
    globalSection = {
      port = cfg.port;
      bind = cfg.bind;
      ping_ip = cfg.pingIp;
      matching_port = cfg.matchingPort;
      allow_register = cfg.allowRegister;
      maintenance_mode = cfg.maintenanceMode;
      enable_paseli = cfg.enablePaseli;
    };
```
This should be RFC42 style and then we don't need to map any options.

File: nixos/modules/services/games/asphyxia-core.nix [hunk 18]
```diff
@@ -0,0 +1,158 @@
+{ config, lib, options, pkgs, ... }:
+
+with lib;
+
+let
+  cfg = config.services.asphyxia-core;
+  opt = options.services.asphyxia-core;
```
Comment (SuperSandro2000) [inline nixos/modules/services/games/asphyxia-core.nix:7]:
```suggestion
```

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]
The commit ``asphyxia service: init`` should be ``nixos/asphyxia: init``

File: pkgs/games/asphyxia-core/default.nix (hunk 4 above)
Comment (arilotter) [reply pkgs/games/asphyxia-core/default.nix:12]: this is required for the mystical hack down below - will move & add a comment.

File: nixos/modules/services/games/asphyxia-core.nix [hunk 19]
```diff
@@ -0,0 +1,158 @@
...
+    CONFIG_DIR=${configDir}
+    echo "starting asphyxia in $CONFIG_DIR"
+    exec -a $CONFIG_DIR/asphyxia-core ${cfg.package}/bin/asphyxia-core -d /var/lib/asphyxia-core
+  '';
+
+  corePlugins = pkgs.fetchgit {
+    url = "https://github.com/asphyxia-core/plugins";
+    rev = "81630a86a299fd278b86b04dd94ece0b36525b2f";
+    sha256 = "sha256-HNtVzlnGbOqXWVbyRqV4sJWOGk9srARPx7vAjmAh4QY=";
+  };
```
[the remark this answers is not in the document]
Comment (arilotter) [reply nixos/modules/services/games/asphyxia-core.nix:40]: I can use fetchFromGithub, but the package has no use for these.
Due to a quirk of how Asphyxia is built, it will look for its own execution path in argv[0] and use that to determine the folder from which to load plugins & config files. Because of this, it's useless to place the plugins in the nix store, since that's read-only - which means users won't be able to put a config file in there, or customize any of the options.

File: nixos/modules/services/games/asphyxia-core.nix (hunk 15 above)
Comment (arilotter) [reply nixos/modules/services/games/asphyxia-core.nix:27]: can do

File: nixos/doc/manual/release-notes/rl-2211.section.md [hunk 20]
```diff
@@ -142,6 +142,8 @@ In addition to numerous new and upgraded packages, this release has the followin
 
 - [Grafana Tempo](https://www.grafana.com/oss/tempo/), a distributed tracing store. Available as [services.tempo](#opt-services.tempo.enable).
 
+- [Asphyxia CORE](https://asphyxia-core.github.io/), an e-amuse emulator for Konami rhythm games. Available as [services.asphyxia-core](#opt-services.asphyxia-core.enable).
```
Comment (SuperSandro2000) [inline nixos/doc/manual/release-notes/rl-2211.section.md:145]: Please rebase and move this to 23.05

**Comment by arilotter** [comment]
I no longer use Asphyxia, and won't be working on this derivation.
